File: Python/AtCoder/old/_ALdS1_10_C_0928.py
def main():
    n = int(input())
    for _ in range(n):
        x = input()
        y = input()
        print(lcs(x, y))


def lcs(x: str, y: str):
    # DP[i][j] = x[:i]とy[:j]のlcs
    # DP[i][j] = もしx[i-1] == y[i-1]ならば、更新可能性あり
    # それ以外は、更新可能性なし
    max_i = len(x)
    max_j = len(y)
    DP = [[0]*(max_j+1) for _ in range(max_i+1)]
    for i in range(max_i+1):
        for j in range(max_j+1):
            if i == 0 or j == 0:
                DP[i][j] = 0
                continue
            if x[i-1] == y[j-1]:
                DP[i][j] = max(DP[i-1][j-1]+1, DP[i][j-1], DP[i-1][j])
            else:
                DP[i][j] = max(DP[i-1][j-1], DP[i][j-1], DP[i-1][j])
    return DP[-1][-1]


main()

File: C++/AtCoder/ふるい/ALDS1_10_C_0928.cpp
// copyright
#include<bits/stdc++.h>
int lcs(std::string x, std::string y) {
    // DP[i][j] = x[:i] y[:j]のlcs
    int max_i = x.length();
    int max_j = y.length();
    int DP[max_i+1][max_j+1];
    for (int i = 0; i <= max_i; i++) {
        for (int j = 0; j <= max_j; j++) {
            DP[i][j] = 0;
        }
    }

    for (int i = 0; i <= max_i; i++) {
        for (int j =  0; j <= max_j; j++) {
            if (i == 0 || j == 0) {
                DP[i][j] = 0;
                continue;
            }
            if (x[i-1] == y[j-1]) {
                DP[i][j] = std::max({DP[i-1][j-1]+1, DP[i][j-1], DP[i-1][j]});
            } else {
                DP[i][j] = std::max({DP[i-1][j-1], DP[i][j-1], DP[i-1][j]});
            }
        }
    }

    return DP[max_i][max_j];
}
int main() {
    int n;
    std::cin >> n;
    for (int i = 0; i < n; i++) {
        std::string x;
        std::string y;
        std::cin >> x >> y;
        int ans;
        ans = lcs(x, y);
        std::cout << ans << std::endl;
    }
}
